File: app/src/main/java/com/tamercad/core/math/Vector3.kt
package com.tamercad.core.math

import kotlin.math.sqrt

data class Vector3(
    var x: Double = 0.0,
    var y: Double = 0.0,
    var z: Double = 0.0
) {
    fun add(v: Vector3): Vector3 = Vector3(x + v.x, y + v.y, z + v.z)

    fun subtract(v: Vector3): Vector3 = Vector3(x - v.x, y - v.y, z - v.z)

    fun multiply(scale: Double): Vector3 = Vector3(x * scale, y * scale, z * scale)

    fun dot(v: Vector3): Double = (x * v.x) + (y * v.y) + (z * v.z)

    fun cross(v: Vector3): Vector3 = Vector3(
        (y * v.z) - (z * v.y),
        (z * v.x) - (x * v.z),
        (x * v.y) - (y * v.x)
    )

    fun length(): Double = sqrt((x * x) + (y * y) + (z * z))

    fun normalize(): Vector3 {
        val len = length()
        if (len == 0.0) return Vector3()
        return Vector3(x / len, y / len, z / len)
    }
}

File: src/math/vector/Vector3.ts
export class Vector3 {

    public x: number;
    public y: number;
    public z: number;


    constructor(
        x = 0,
        y = 0,
        z = 0
    ) {
        this.x = x;
        this.y = y;
        this.z = z;
    }


    add(v: Vector3): Vector3 {

        return new Vector3(
            this.x + v.x,
            this.y + v.y,
            this.z + v.z
        );
    }


    subtract(v: Vector3): Vector3 {

        return new Vector3(
            this.x - v.x,
            this.y - v.y,
            this.z - v.z
        );
    }


    multiply(scale:number): Vector3 {

        return new Vector3(
            this.x * scale,
            this.y * scale,
            this.z * scale
        );
    }


    dot(v:Vector3):number {

        return (
            this.x * v.x +
            this.y * v.y +
            this.z * v.z
        );
    }


    cross(v:Vector3):Vector3 {

        return new Vector3(
            this.y * v.z - this.z * v.y,
            this.z * v.x - this.x * v.z,
            this.x * v.y - this.y * v.x
        );
    }


    length():number {

        return Math.sqrt(
            this.x*this.x +
            this.y*this.y +
            this.z*this.z
        );
    }


    normalize():Vector3 {

        const len = this.length();

        if(len === 0)
            return new Vector3();


        return new Vector3(
            this.x / len,
            this.y / len,
            this.z / len
        );
    }
}